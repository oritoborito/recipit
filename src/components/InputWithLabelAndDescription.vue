<script setup lang="ts">
type ComponentType = 'input' | 'textarea' | 'select'

const props = withDefaults(
    defineProps<{
      title?: string
      label: string
      description: string
      value: string | number
      renderAsComponentType?: ComponentType
      placeholder?: string
      inputType?: 'text' | 'number'
      pattern?: string
    }>(),
    {
      inputType: 'text',
      renderAsComponentType: 'input'
    }
)

const inputId = Math.random().toString(36).slice(2)

const isRenderedAsType = (type: ComponentType) => {
  return props.renderAsComponentType === type
}
const getValueIfComponentIsOfType = (type: ComponentType, value: string | number | undefined) => {
  return isRenderedAsType(type) ? value : undefined
}
</script>

<template>
  <div class="container">
    <h2 v-if="title">{{ title }}</h2>
    <div>
      <label :for="inputId">{{ label }}</label>
      <span>{{ description }}</span>
      <component
          :is="renderAsComponentType"
          :id="inputId"
          :pattern="getValueIfComponentIsOfType('input', pattern)"
          :type="getValueIfComponentIsOfType('input', inputType)"
          :rows="getValueIfComponentIsOfType('textarea',10)"
          :placeholder="placeholder"
          :value="value"
          @input="$emit('update:value', $event.target.value)"
      >
        <slot/>
      </component>
  </div>
  </div>
</template>

<style scoped>
.container {
  display: flex;
  flex-direction: column;
  gap: 1rem;
}

div {
  display: grid;
  width: fit-content;
}

span {
  font-size: 0.8rem;
}

textarea, input, select {
  border: var(--border);
  border-radius: var(--radius);
  padding: 0.5rem;
  font-family: inherit;
  background-color: transparent;
  color: var(--color-text);
  font-size: 1rem;
}


</style>
