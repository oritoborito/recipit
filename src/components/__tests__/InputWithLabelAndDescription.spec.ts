import { describe, it, expect } from 'vitest'

import { mount } from '@vue/test-utils'
import InputWithLabelAndDescription from "../InputWithLabelAndDescription.vue";

describe('InputWithLabelAndDescription', () => {
  it('renders requried props', () => {
    const wrapper = mount(InputWithLabelAndDescription, {
      props: {
        label: 'label',
        description: 'description',
        value: 'value',
      }
    })
    expect(wrapper.find('label').text()).toBe('label')
    expect(wrapper.find('input').element.value).toBe('value')
    expect(wrapper.find('span').text()).toBe('description')
  })
})
