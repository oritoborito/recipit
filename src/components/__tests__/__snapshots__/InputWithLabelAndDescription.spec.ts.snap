// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputWithLabelAndDescription > renders properly 1`] = `
"<div data-v-062c9662=\\"\\" class=\\"container\\">
  <!--v-if-->
  <div data-v-062c9662=\\"\\"><label data-v-062c9662=\\"\\" for=\\"armh77yi8ti\\">label</label><span data-v-062c9662=\\"\\">description</span><input data-v-062c9662=\\"\\" id=\\"armh77yi8ti\\" type=\\"text\\"></div>
</div>"
`;
